### PR TITLE
Vehicle water update + minor crafting refactoring

### PR DESCRIPTION
**Pull request by Coolthulhu**
`map::use_charges` and `inventory::from_map` refactored.
The former changes crafting behavior slightly, but it shouldn't actually change item costs. Old version would go like this:

```
33333
32223
32123
32223
33333
```

First going through 1, then going through all the 2 (starting in upper-left corner) AND the 1, then all the 3 (starting in upper-left) AND the 2 and 1 etc.
New behavior uses `closest_points_first`, which means it will go over each point at most once, but more importantly it uses one loop instead of 3 so it looks much cleaner.

---

Changed water cannon from clean water to dirty water. Changed default ammo type of "water" ammo to dirty water.
Changed tanks in fire engine from clean water to dirty water.

---

Added a new vehicle water purifier part. It can be used to purify water like FOODCO, but also to purify water in vehicle's tanks (only when there's enough batteries, dirty water and the vehicle has enough clean water capacity). It is installed from a regular handheld purifier at mechanics 4 and occupies center part.

---

Allowed kitchens and faucets to provide dirty water for crafting. Not for drinking/refilling though - the `item::poison` variable isn't stored anywhere, so it should probably be recreated with maximal value (one from toilet water or sewage), which basically means it shouldn't be consumed.

### CODE REVIEW

File: src/pickup.cpp [hunk 1]
```diff
@@ -24,181 +24,203 @@ int Pickup::interact_with_vehicle( vehicle *veh, const tripoint &pos, int veh_ro
...
-        }
+    if( veh == nullptr ) {
+        return -1;
+    }
 
-        if( from_vehicle ) {
-            menu_items.push_back(_("Get items"));
-            options_message.push_back(uimenu_entry(_("Get items"), 'g'));
-        }
+    k_part = veh->part_with_feature(veh_root_part, "KITCHEN");
```
Comment (BevapDin) [inline src/pickup.cpp]: What about moving the declaration of the *_part variables here where they are initialized? Could also make them const and would avoid the completely useless initialization to 0, -1 or whatever, the value is overridden anyway.

File: src/map.cpp [hunk 2]
```diff
@@ -4741,6 +4741,7 @@ std::list<item> map::use_amount( const tripoint &origin, const int range, const
 template <typename Stack>
 std::list<item> use_charges_from_stack( Stack stack, const itype_id type, long &quantity)
 {
+    // TODO: Fix the bug with this function consuming liquid off the ground
     std::list<item> ret;
     for( auto a = stack.begin(); a != stack.end() && quantity > 0; ) {
         if( a->use_charges(type, quantity, ret) ) {
```
Comment (BevapDin) [inline src/map.cpp]: What about:
```C++
if(a->made_of(LIQUID)) {
    ++a;
    continue;
}
```
to skip liquids directly on the ground (the TODO)?

Comment (Barhandar) [inline src/map.cpp]: Do the inherent liquids that don't show up in 'e', like water pump's or water tile's, count for your check?

Comment (BevapDin) [inline src/map.cpp]: > Do the inherent liquids that don't show up in 'e', like water pump's or water tile's, count for your check?

This check is only for items on the square. Furniture/terrain that is able to produce water does not contain water items, so it won't apply here.

**Pull request merged**